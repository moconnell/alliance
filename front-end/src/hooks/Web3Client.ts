import Web3Modal from "web3modal";
import WalletConnectProvider from "@walletconnect/web3-provider";
import { useEffect, useReducer, useCallback, useMemo } from "react";
import { ethers } from "ethers";
import { Web3ProviderState, web3InitialState, web3Reducer } from "../reducers";

const getWeb3Modal = () => {
  if (typeof window === "undefined") return null;

  const {
    REACT_APP_WEB3_CACHE_PROVIDER,
    REACT_APP_WEB3_INFURA_ID,
    REACT_APP_WEB3_NETWORK,
  } = process.env;

  return new Web3Modal({
    network: REACT_APP_WEB3_NETWORK,
    cacheProvider: REACT_APP_WEB3_CACHE_PROVIDER,
    providerOptions: {
      walletconnect: {
        package: WalletConnectProvider,
        options: {
          infuraId: REACT_APP_WEB3_INFURA_ID,
        },
      },
    },
  });
};

export type Web3Client = Web3ProviderState & {
  connect: () => Promise<void>;
  disconnect: () => Promise<void>;
};

export const useWeb3 = () => {
  const [state, dispatch] = useReducer(web3Reducer, web3InitialState);
  const { provider, web3Provider, address, network, signer } = state;
  const web3Modal = useMemo(() => getWeb3Modal(), []);

  const connect = useCallback(async () => {
    try {
      if (!web3Modal) {
        console.error("Cannot connect: no Web3Modal");
        return;
      }

      const provider = await web3Modal.connect();
      const web3Provider = new ethers.providers.Web3Provider(provider);
      const signer = web3Provider.getSigner();
      const address = await signer.getAddress();
      const network = await web3Provider.getNetwork();

      dispatch({
        type: "SET_WEB3_PROVIDER",
        provider,
        web3Provider,
        address,
        network,
        signer,
      });
    } catch (e) {
      console.log("connect error", e);
    }
  }, [web3Modal]);

  const disconnect = useCallback(async () => {
    if (web3Modal) {
      web3Modal.clearCachedProvider();
      if (provider?.disconnect && typeof provider.disconnect === "function") {
        await provider.disconnect();
      }
      dispatch({
        type: "RESET_WEB3_PROVIDER",
      });
    } else {
      console.error("No Web3Modal");
    }
  }, [provider, web3Modal]);

  // Auto connect to the cached provider
  useEffect(() => {
    if (web3Modal && web3Modal.cachedProvider) {
      connect();
    }
  }, [connect, web3Modal]);

  // EIP-1193 events
  useEffect(() => {
    if (provider?.on) {
      const handleAccountsChanged = (accounts: string[]) => {
        dispatch({
          type: "SET_ADDRESS",
          address: accounts[0],
        });
      };

      // https://docs.ethers.io/v5/concepts/best-practices/#best-practices--network-changes
      const handleChainChanged = (_hexChainId: string) => {
        if (typeof window !== "undefined") {
          console.log("switched to chain...", _hexChainId);
          window.location.reload();
        } else {
          console.log("window is undefined");
        }
      };

      const handleDisconnect = (error: { code: number; message: string }) => {
        // eslint-disable-next-line no-console
        console.log("disconnect", error);
        disconnect();
      };

      provider.on("accountsChanged", handleAccountsChanged);
      provider.on("chainChanged", handleChainChanged);
      provider.on("disconnect", handleDisconnect);

      // Subscription Cleanup
      return () => {
        if (provider.removeListener) {
          provider.removeListener("accountsChanged", handleAccountsChanged);
          provider.removeListener("chainChanged", handleChainChanged);
          provider.removeListener("disconnect", handleDisconnect);
        }
      };
    }
  }, [provider, disconnect]);

  return {
    provider,
    web3Provider,
    address,
    network,
    signer,
    connect,
    disconnect,
  } as Web3Client;
};
