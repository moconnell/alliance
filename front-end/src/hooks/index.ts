export type { Web3Client } from "./Web3Client";
export { useWeb3 } from "./Web3Client";
export { useCalendar } from "./useCalendar";
