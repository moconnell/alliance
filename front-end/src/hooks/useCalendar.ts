import { useCallback, useEffect, useState } from "react";

import { useWeb3Context } from "../context/Web3Context";
import assert from "assert";
import {
  Calendar,
  CalendarFactory,
  CalendarFactory__factory,
  Calendar__factory,
} from "../../typechain-types";
import DaysOfWeek from "../types/daysOfWeek";
import {
  assertValid,
  compare,
  fromTotalMins,
  totalMinutes,
} from "../utils/timeUtils";
import Meeting from "../types/meeting";
import ProfileInfo from "../types/profileInfo";
import AvailabilityInfo from "../types/availabilityInfo";
import { DateTime } from "luxon";

const NULL_ADDRESS = "0x0000000000000000000000000000000000000000";

const toString = (value: any) => (value ? String(value) : "");

export const useCalendar = () => {
  const { address, signer, web3Provider } = useWeb3Context();
  const [calendarFactory, setCalendarFactory] = useState<CalendarFactory>();
  const [calendar, setCalendar] = useState<Calendar>();
  const [availability, setAvailability] = useState<AvailabilityInfo>();
  const [profile, setProfile] = useState<ProfileInfo>();

  const { REACT_APP_WEB3_CONTRACT_FACTORY_ADDRESS } = process.env;
  assert(
    REACT_APP_WEB3_CONTRACT_FACTORY_ADDRESS,
    "REACT_APP_WEB3_CONTRACT_FACTORY_ADDRESS not set"
  );

  useEffect(() => {
    function getCalendarFactory() {
      if (!web3Provider || !address || !signer) return undefined;
      return CalendarFactory__factory.connect(
        REACT_APP_WEB3_CONTRACT_FACTORY_ADDRESS!,
        signer
      );
    }

    const fac = getCalendarFactory();
    if (fac) setCalendarFactory(fac);
  }, [REACT_APP_WEB3_CONTRACT_FACTORY_ADDRESS, address, signer, web3Provider]);

  useEffect(() => {
    async function getCalendar() {
      if (!address || !calendarFactory || !web3Provider) return;
      console.log(calendarFactory);
      const calendarAddress = await calendarFactory.userToCalendar(address);
      if (calendarAddress === NULL_ADDRESS) {
        console.log(`wallet ${address} has no saved calendar`);
        return;
      }

      console.log(`fetching calendar at ${calendarAddress}`);

      const cal = Calendar__factory.connect(calendarAddress, web3Provider);
      setCalendar(cal);
    }

    getCalendar();
  }, [address, calendarFactory, web3Provider]);

  useEffect(() => {
    async function loadProfileAvailability() {
      if (!calendar) return;

      const newProfile = await calendar.profile();
      setProfile(newProfile);

      let newAvailability = await calendar.availability();
      const from = fromTotalMins(newAvailability.earliestStartMinutes);
      const to = fromTotalMins(
        newAvailability.earliestStartMinutes + newAvailability.minutesAvailable
      );
      setAvailability({ ...newAvailability, from, to });
    }

    loadProfileAvailability();
  }, [calendar]);

  useEffect(() => {
    const createCalendar = async () => {
      if (
        calendar ||
        !calendarFactory ||
        !web3Provider ||
        !profile ||
        !availability
      )
        return;

      assert(profile.email, "No email");
      assertValid(availability.from);
      assertValid(availability.to);
      assert(
        compare(availability.from!, availability.to!) < 0,
        "from after to"
      );

      const p = {
        description: toString(profile.description),
        email: profile.email!,
        picture: toString(profile.picture),
        url: toString(profile.url),
        username: toString(profile.username),
      };

      const a = {
        availableDays: availability.availableDays ?? DaysOfWeek.None,
        earliestStartMinutes: totalMinutes(availability.from!),
        minutesAvailable:
          totalMinutes(availability.to!) - totalMinutes(availability.from!),
        location: availability.location!,
        timeZone: availability.timeZone!,
      };

      console.log("creating calendar now ... ", p, a);

      const tx = await calendarFactory.createCalendar(p, a);
      const receipt = await tx.wait();
      const txEvent = receipt.events?.[0]?.args;
      const calendarAddr = txEvent?.calendar;
      const cal = Calendar__factory.connect(calendarAddr, web3Provider);

      console.log("calendar created ok", cal);

      setCalendar(cal);
    };

    createCalendar();
  }, [
    profile,
    availability,
    calendar,
    calendarFactory,
    web3Provider,
    address,
    REACT_APP_WEB3_CONTRACT_FACTORY_ADDRESS,
  ]);

  const getMeetings = useCallback(
    async (date: Date) => {
      if (!calendar) return undefined;

      console.log(`getMeetings(date: ${date})`);

      const year = date.getFullYear();
      const month = date.getMonth() + 1;
      const day = date.getDate();
      const meetingsOnDate = await calendar.getMeetings(year, month, day);

      return meetingsOnDate.map(
        ({ attendee, startMinutes, durationMinutes }) => {
          const hour = Math.floor(startMinutes / 60);
          const minute = startMinutes % 60;
          const from = DateTime.fromObject(
            {
              year,
              month,
              day,
              hour,
              minute,
            },
            {
              zone: availability?.timeZone,
            }
          ).setZone();
          const to = from.plus({ minutes: durationMinutes });

          return {
            date: from,
            endDate: to,
            attendee: attendee,
            description: `${durationMinutes} minute meeting`,
          } as Meeting;
        }
      );
    },
    [availability?.timeZone, calendar]
  );

  const getProfileAvailability = useCallback(
    async (calendarAddress: string) => {
      if (!web3Provider) return undefined;

      console.log(
        `getProfileAvailability(calendarAddress: ${calendarAddress})`
      );

      const otherCalendar = Calendar__factory.connect(
        calendarAddress,
        web3Provider
      );

      const profileAvailability: [ProfileInfo, AvailabilityInfo] = [
        await otherCalendar.profile(),
        await otherCalendar.availability(),
      ];

      return profileAvailability;
    },
    [web3Provider]
  );

  const getAvailableTimes = useCallback(
    async (calendarAddress: string, date: Date, durationMinutes: number) => {
      if (!web3Provider) return undefined;

      console.log(
        `getAvailableTimes(calendarAddress: ${calendarAddress}, date: ${date}, duration: ${durationMinutes})`
      );

      const otherCalendar = Calendar__factory.connect(
        calendarAddress,
        web3Provider
      );

      const year = date.getFullYear();
      const month = date.getMonth() + 1;
      const day = date.getDate();
      const { times, timeZone } = await otherCalendar.getAvailableTimes(
        year,
        month,
        day,
        durationMinutes
      );

      return times
        .filter((startMinutes) => startMinutes > 0)
        .map((startMinutes) => {
          const hour = Math.floor(startMinutes / 60),
            minute = startMinutes % 60;

          return DateTime.fromObject(
            { year, month, day, hour, minute },
            { zone: timeZone }
          );
        });
    },
    [web3Provider]
  );

  const bookMeeting = useCallback(
    async (calendarAddress: string, at: DateTime, durationMinutes: number) => {
      if (!signer) return;

      console.log(
        `bookMeeting(calendarAddress: ${calendarAddress}, at: ${at}, duration: ${durationMinutes})`
      );

      const otherCalendar = Calendar__factory.connect(calendarAddress, signer);

      await otherCalendar.bookMeeting(
        at.year,
        at.month,
        at.day,
        at.hour,
        at.minute,
        durationMinutes
      );
    },
    [signer]
  );

  const setProfileAvailability = (info: ProfileInfo & AvailabilityInfo) => {
    setProfile(info);
    setAvailability(info);
  };

  return {
    address,
    availability,
    bookMeeting,
    getAvailableTimes,
    getMeetings,
    getProfileAvailability,
    hasCalendar: calendar !== undefined,
    profile,
    setProfileAvailability,
  };
};
