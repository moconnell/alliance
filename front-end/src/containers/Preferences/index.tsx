import { useEffect, useMemo } from "react";
import { Navigate } from "react-router-dom";
import { FormProvider, useForm } from "react-hook-form";
import Input from "../../components/Input";
import { useCalendar } from "../../hooks";
import DaysOfWeek from "../../types/daysOfWeek";
import Time from "../../types/time";

const Preferences = () => {
  const { address, availability, profile, setProfileAvailability } = useCalendar();

  const defaultValues = useMemo(() => {
    const defaultValues = { ...profile, ...availability };
    if (!defaultValues.availableDays)
      defaultValues.availableDays = DaysOfWeek.MonFri;
    if (!defaultValues.from) defaultValues.from = { hours: 9, minutes: 0 } as Time;
    if (!defaultValues.to) defaultValues.to = { hours: 12, minutes: 0 } as Time;
    if (!defaultValues.timeZone)
      defaultValues.timeZone = Intl.DateTimeFormat().resolvedOptions().timeZone;
    if (!defaultValues.location)
      defaultValues.location = defaultValues.timeZone.split("/")[1];

    return defaultValues;
  }, [availability, profile]);

  const methods = useForm({
    defaultValues,
    mode: "onChange",
  });

  const { handleSubmit, reset } = methods;

  useEffect(() => {
    reset(defaultValues);
  }, [defaultValues, reset]);

  if (!address) {
    return <Navigate to="/" />;
  }

  return (
    <FormProvider {...methods}>
      <form aria-label="form" onSubmit={handleSubmit(setProfileAvailability)}>
        <Input />
      </form>
    </FormProvider>
  );
};

export default Preferences;
