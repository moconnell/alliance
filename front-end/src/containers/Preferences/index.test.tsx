import { render } from "@testing-library/react";
import Preferences from "./index";
import { createMemoryHistory } from "history";
import { useCalendar } from "../../hooks";
import { Router, Route, Routes } from "react-router-dom";
import { Fragment } from "react";
import { web3InitialState } from "../../reducers";

const PROFILE_PATH = "/profile";
const ROOT_PATH = "/";
const MOCK_INPUT = "((Input))";
const MOCK_HOME = "((Home))";

jest.mock("react-hook-form", () => {
  return {
    ...jest.requireActual("react-hook-form"),
    useForm: () => {
      return { handleSubmit: jest.fn(), reset: jest.fn() };
    },
  };
});
jest.mock("../../hooks");
jest.mock("../../components/Input", () => () => MOCK_INPUT);

describe("<Preferences />", () => {
  const mockSetProfileAvailability = jest.fn();

  afterEach(() => jest.resetAllMocks());
  afterAll(() => jest.restoreAllMocks());

  const testData: [string, string, string?][] = [
    [
       "render correctly",
       PROFILE_PATH,
       "0xB7A5bd0345EF1Cc5E66bf61BdeC17D2461fBd968",
    ],
    [
       "redirect to home page",
       ROOT_PATH,
    ],
  ];

  testData.forEach(([ name, expectedRoute, address ]) =>
    it(`should ${name}`, () => {
      (useCalendar as jest.Mock).mockImplementation(() => {
        return {
          address,
          availablility: {},
          profile: {},
          setProfileAvailability: mockSetProfileAvailability,
        };
      });
      
      const history = createMemoryHistory();
      history.push(PROFILE_PATH);

      const { asFragment } = render(
        <Router location={history.location} navigator={history}>
          <Routes>
            <Route path={PROFILE_PATH} element={<Preferences />} />
            <Route
              path={ROOT_PATH}
              element={<Fragment>{MOCK_HOME}</Fragment>}
            />
            <Route element={<Fragment>No route match</Fragment>} />
          </Routes>
        </Router>
      );

      expect(history.location.pathname).toBe(expectedRoute);
      expect(asFragment()).toMatchSnapshot();
    })
  );
});
