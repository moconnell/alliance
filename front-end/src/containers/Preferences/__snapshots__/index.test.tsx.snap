// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Preferences /> should redirect to home page 1`] = `<DocumentFragment />`;

exports[`<Preferences /> should render correctly 1`] = `
<DocumentFragment>
  <form
    aria-label="form"
  >
    ((Input))
  </form>
</DocumentFragment>
`;
